forc-test/src/lib.rs

```diff
@@ -362,12 +372,67 @@ impl BuiltTests {
     }
 }
 
+/// Build all contracts in the given buld plan without tests.
+fn build_contracts_without_tests(
+    opts: &Opts,
+    build_plan: &pkg::BuildPlan,
+) -> Vec<(pkg::Pinned, anyhow::Result<BuiltPackage>)> {
+    let manifest_map = build_plan.manifest_map();
+    build_plan
+        .member_pinned_pkgs()
+        .map(|pinned_pkg| {
+            let pkg_manifest = manifest_map
+                .get(&pinned_pkg.id())
+                .expect("missing manifest for member to test");
+            (pinned_pkg, pkg_manifest)
+        })
+        .filter(|(_, pkg_manifest)| matches!(pkg_manifest.program_type(), Ok(TreeType::Contract)))
+        .map(|(pinned_pkg, pkg_manifest)| {
+            let pkg_path = pkg_manifest.dir();
+            let build_opts_without_tests = opts
+                .clone()
+                .patch_opts(pkg_path)
+                .into_build_opts()
+                .include_tests(false);
+            let built_pkg =
+                pkg::build_with_options(build_opts_without_tests).and_then(|pkg| pkg.expect_pkg());
+            (pinned_pkg, built_pkg)
+        })
+        .collect()
+}
```

Comment:
I wanted to return an iterator over the elements rather than a `Vec` but couldn't manage to get pass the compiler due to strange lifetime complains. I ended up collecting them this dynamically allocates a `Vec` but if I could return an iterator over `BuiltPackage` references, I would need need to clone them in `build()` as I need the owned version to store in `built_contracts` since I would `clone` all `BuiltPackage` references, I feel like performance wise these two options should be the same but wanted to hear your thoughts on this @mitchmindtree

Reply:
In this case there likely won't be a noticeable performance impact as when using this function, we end up storing all of the `BuiltPackage`s either way. However, from an API design perspective it's almost always better to return an Iterator if feasible as it gives the function user more control over their costs.

> couldn't manage to get pass the compiler due to strange lifetime complains

Sometimes when there are multiple references as function inputs, Rust can't safely infer the correct lifetime for the output, so using `'_` in something like this doesn't work:

```rust
fn build_contracts_without_tests(
    opts: &Opts,
    build_plan: &pkg::BuildPlan,
) -> impl '_ + Iterator<Item = (pkg::Pinned, anyhow::Result<BuiltPackage>)> {
```

You can work around this by explicitly naming the lifetime, so it's clear to Rust what inputs the lifetime is bound to, e.g.

```rust
fn build_contracts_without_tests<'a>(
    opts: &'a Opts,
    build_plan: &'a pkg::BuildPlan,
) -> impl 'a + Iterator<Item = (pkg::Pinned, anyhow::Result<BuiltPackage>)> {
```